pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,342 @@
+	}
+}
+
+func (rc reconcileCounter) reconcileWith(request reconcile.Request) {
+	if rc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
+		}
+	}
+	rc.counterVec.WithLabelValues(rc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type reconcileFailureCounter struct {
+	groupVersionKind string
+	namespaced       bool
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileFailureCounterFor(gvk schema.GroupVersionKind, scope meta.RESTScopeName) reconcileFailureCounter {
+	if scope == meta.RESTScopeNameNamespace {
+		return reconcileFailureCounter{groupVersionKind: gvkString(gvk), namespaced: true, counterVec: reconcileFailure}
+	} else {
+		return reconcileFailureCounter{groupVersionKind: gvkString(gvk), namespaced: false, counterVec: reconcileFailure}
+	}
+}
+
+func (rfc reconcileFailureCounter) failureWith(request reconcile.Request) {
+	if rfc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
+		}
+	}
+	rfc.counterVec.WithLabelValues(rfc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type objectRecorder struct {
+	groupVersionKind string
+	gaugeVec         *prometheus.GaugeVec
+}
+
+func objectRecorderFor(gvk schema.GroupVersionKind) objectRecorder {
+	return objectRecorder{groupVersionKind: gvkString(gvk), gaugeVec: managedObjectsRecord}
+}
+
+func (or objectRecorder) Set(namespace, name string, num float64) {
+	if namespace == ClusterScoped {
+		or.gaugeVec.WithLabelValues(or.groupVersionKind, "", name).Set(num)
+	} else {
+		or.gaugeVec.WithLabelValues(or.groupVersionKind, namespace, name).Set(num)
+	}
+}
+
+var globalObjectTracker *objectTracker = newObjectTracker()
+
+type objectTracker struct {
+	mu         sync.Mutex
+	mgr        manager.Manager
+	trackedGVK map[schema.GroupVersionKind]*gvkTracker
+}
+
+// Call this before kubectl.Apply
+func (ot *objectTracker) addIfNotPresent(object *manifest.Object) error {
+	ot.mu.Lock()
+	defer ot.mu.Unlock()
+
+	gvk, ns, name := object.GroupVersionKind(),
+		object.UnstructuredObject().GetNamespace(),
+		object.UnstructuredObject().GetName()
+
+	namespaced, err := namespaced(ot.mgr, gvk)
+	if err != nil {
+		return err
+	}
+
+	if namespaced {
+		if len(ns) == 0 {
+			ns = DefaultNamespace
+		}
+	} else {
+		ns = ClusterScoped
+	}
+
+	if gvkt, ok := ot.trackedGVK[gvk]; ok {
+		if gvkt.has(ns, name) {
+			return nil
+		} else {
+			gvkt.insert(ns, name)
+			return nil
+		}
+	}
+
+	ot.trackedGVK[gvk] = newGVKTracker(ot.mgr, object.UnstructuredObject(), namespaced)
+	ot.trackedGVK[gvk].insert(ns, name)
+
+	// addIfNotPresent is called at Reconcler.reconcileExists,
+	// so Controller & Manager is already running
+	ot.trackedGVK[gvk].start()
+
+	return nil
+}
+
+func newObjectTracker() *objectTracker {
+	ot := &objectTracker{}
+	ot.mgr = nil
+	ot.trackedGVK = make(map[schema.GroupVersionKind]*gvkTracker)
+
+	return ot
+}
+
+type gvkTracker struct {
```

Comment:
This is impressive, but a lot of work.  What do we think this metric will be used for?  Should we track a counter of the number of applies instead (which might be easier, and give similar results, depending on what we want it for)?